cirq/google/serializable_device.py

```diff
@@ -0,0 +1,238 @@
+            gate_set: A SerializableGateSet that can translate the gate_ids
+                into cirq Gates.
+        """
+
+        # Store target sets, since they are refered to by name later
+        allowed_targets: Dict[str, Set[Tuple['cirq.Qid', ...]]] = {}
+        permutation_ids: Set[str] = set()
+        for ts in proto.valid_targets:
+            allowed_targets[ts.name] = cls._create_target_set(ts)
+            if ts.target_ordering == v2.device_pb2.TargetSet.SUBSET_PERMUTATION:
+                permutation_ids.add(ts.name)
+
+        # Store gate definitions from proto
+        gate_defs: Dict[str, v2.device_pb2.GateDefinition] = {}
+        for gs in proto.valid_gate_sets:
+            for gate_def in gs.valid_gates:
+                gate_defs[gate_def.id] = gate_def
+
+        # Loop through serializers and create dictionaries with keys as
+        # the python type of the cirq Gate and duration/targets from the proto
+        durations: Dict[Type['cirq.Gate'], Duration] = {}
+        target_sets: Dict[Type['cirq.Gate'], Set[
+            Tuple['cirq.Qid', ...]]] = dict()
+        permutation_gates: Set[Type['cirq.Gate']] = set()
+        for gate_type in gate_set.supported_gate_types():
+            for serializer in gate_set.serializers[gate_type]:
+                gate_id = serializer.serialized_gate_id
+                if gate_id not in gate_defs:
+                    raise ValueError(f'Serializer has {gate_id} which is not '
+                                     'supported by the device specification')
+                gate_ts = gate_defs[gate_id].valid_targets
+                which_are_permutations = [t in permutation_ids for t in gate_ts]
+                if any(which_are_permutations):
+                    if not all(which_are_permutations):
+                        msg = f'Id {gate_id} in {gate_set.gate_set_name} ' \
+                            ' mixes SUBSET_PERMUTATION with other types which' \
+                            ' is not currently allowed.'
+                        raise NotImplementedError(msg)
+                    permutation_gates.add(gate_type)
+                gate_picos = gate_defs[gate_id].gate_duration_picos
+                durations[gate_type] = Duration(picos=gate_picos)
+                if gate_type not in target_sets:
+                    target_sets[gate_type] = set()
+                for target_set_name in gate_defs[gate_id].valid_targets:
+                    target_sets[gate_type] |= allowed_targets[target_set_name]
+
+        return SerializableDevice(
+            qubits=SerializableDevice._qubits_from_ids(proto.valid_qubits),
+            durations=durations,
+            target_sets=target_sets,
+            permutation_gates=permutation_gates,
+        )
+
+    @staticmethod
+    def _qid_from_str(id_str: str) -> 'cirq.Qid':
+        try:
+            return devices.GridQubit.from_proto_id(id_str)
+        except ValueError:
+            return ops.NamedQubit(id_str)
```

Comment:
This fallback behavior can create some very confusing issues. For example, suppose I put a LineQubit in somewhere and it gets turned into "1". When it gets parsed here, I'll get a NamedQubit("1") != LineQubit(1) out.

Reply:
now you understand why my json schema is so verbose :)

Reply:
I'm okay with this behavior.  LineQubits are probably not going to be serialized as part of Google's devices (and this is in the google directory).  This is further reinforced by the fact that LineQubit's don't have proto_id() or from_proto_id() functions.

I did put pydoc with description that explicitly only supports GridQubit and NamedQubit.